rfcs/0003-multi-blockchain-support.md

```diff
@@ -0,0 +1,534 @@
+smart contracts (if supported by the respective chain), handling reorgs (if
+applicable) and triggering off of blockchain-specific features like events or
+transactions.
+
+## Urgency
+
+Since this is a big feature and one that requires a substantial refactoring of
+the existing codebase, and _may_ introduce breaking changes, it would be good to
+start working on this as soon as possible.
+
+## Terminology
+
+The feature proposed in this RFC is referred to as **multi-blockchain support**.
+Other terms that are introduced in this RFC or are relevant when talking about
+it are listed below.
+
+- **Chain** - a Blockchain technology or similar, such as Ethereum or Substrate.
+- **Network** - a specific instance of a _chain_.
+- **Network subgraph** - a built-in subgraph that indexes data intrinsic to a
+  _network_, such as its blocks, transactions, accounts and balances.
+- **Trigger** - a _chain_ event or similar that can be observed on a _network_
+  and that can be used to trigger indexing work via a subgraph mapping.
+  Examples: blocks, transactions, contract method calls, logs/events.
+
+## Detailed Design
+
+Multi-blockchain support touches almost every aspect of The Graph, including:
+
+- Subgraph manifests and data sources
+- Code generation
+- Mapping APIs and types
+- Subgraph indexing
+- Storage
+
+This section specifies how all of these components are updated to support other
+chains and what the integration of additional chains looks like. With this
+feature implemented, subgraph developers will be able to define new types of
+data sources for supported chains. Indexers will be able to configure which of
+these chains they want to enable.
+
+### Subgraph Manifest
+
+The structure of the subgraph manifest is already desigend for being extended
+with new types of data sources. An Ethereum data source is currently defined with
+
+```yaml
+kind: ethereum/contract
+name: Gravity
+source:
+  abi: Gravity
+  network: mainnet
+  address: '2E645469f354BB4F5c8a05B3b30A929361cf77eC'
+  startBlock: 5000000
+mapping:
+  kind: wasm/assemblyscript
+  apiVersion: 0.0.3
+  eventHandlers:
+  - event: NewGravatar(address,address,uint256)
+    handler: handleNewGravatar
+  callHandlers:
+  - function: createGravatar(string,string)
+    handler: handleCreateGravatar
+  blockHandlers:
+  - handler: handleBlock
+```
+
+Multi-blockchain opens this up to other chains by allowing data sources to have
+`kind` values other than `ethereum/contract`. Support for individual chains will
+require separate RFCs, each of which will introduce one or more data source
+types with their own `kind` values such as e.g. `substrate/contract`.
+
+One change that simplifies working with ABIs and makes the code generated for them
+available across data sources is to move ABIs to the top level of the manifest
+and mark them with a `kind` as well:
+
+```yaml
+kind: subgraph:1.0.0
+schema:
+  file: ./schema.graphql
+abis:
```

Comment:
This enshrines the concept of "ABIs" in our protocol. Are we sure that all blockchains have an equivalent concept?

Reply:
I.e., Might some chains use a grpc IDL to define what methods could be called, and if so, should we use a more generic name here?

Reply:
I'm fine with that. I wonder what a better term would be... "bindings"?

Reply:
I guess it depends on what other things end up going under here. "bindings" could work. "interface-descriptions" also seems appropriate as typically external API types/methods are specified in an IDL.

Reply:
Even after looking at a number of other chains, what else we would put in here remains a little unclear. We'll find a better name than `abis`.